subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c

```diff
@@ -0,0 +1,241 @@
+			return -ENOMEM;
+		}
+	} else {
+		if (params->security == 2)
+			ssid->key_mgmt = WPA_KEY_MGMT_PSK_SHA256;
+		else
+			ssid->key_mgmt = WPA_KEY_MGMT_PSK;
+
+		str_clear_free(ssid->passphrase);
+		ssid->passphrase = dup_binstr(params->psk, params->psk_length);
+
+		if (ssid->passphrase == NULL) {
+			LOG_ERR("%s:Failed to copy passphrase\n", __func__);
+			return -ENOMEM;
+		}
+	}
+
+	wpa_config_update_psk(ssid);
+
+	return 0;
+}
+
+static inline int add_network(struct wpa_supplicant *wpa_s, struct wifi_connect_req_params *params)
+{
+	int ret = 0;
+	struct wpa_ssid *network = wpa_supplicant_add_network(wpa_s);
+
+	if (network == NULL) {
+		return -EINVAL;
+	}
+
+	network->ssid = os_zalloc(sizeof(u8) * WIFI_SSID_MAX_LEN);
+
+	if (network->ssid == NULL) {
+		return -ENOMEM;
+	}
+
+	memcpy(network->ssid, params->ssid, params->ssid_length);
+	network->ssid_len = params->ssid_length;
+	network->disabled = 1;
+	network->key_mgmt = WPA_KEY_MGMT_NONE;
+	network->ieee80211w = params->mfp;
+
+	ret = channel_to_freq(params, network);
+	if (ret) {
+		return ret;
+	}
+	wpa_s->conf->filter_ssids = 1;
+	wpa_s->conf->ap_scan = 1;
+
+	ret = copy_psk(params, network);
+	if (ret) {
+		return ret;
+	}
+	wpa_supplicant_enable_network(wpa_s, network);
+	return 0;
+}
+
+static void add_stored_network(void *cb_arg, const char *ssid, size_t ssid_len)
+{
+	int ret = 0;
+	struct wpa_supplicant *wpa_s = cb_arg;
+	struct wifi_connect_req_params params = { 0 };
+	struct wifi_credentials_personal creds;
+
+	/* load stored data */
+	ret = wifi_credentials_get_by_ssid_personal_struct(ssid, ssid_len, &creds);
+	if (ret) {
+		LOG_ERR("Loading WiFi credentials failed for SSID [%.*s], len: %d, err: %d",
+			ssid_len, ssid,  ssid_len, ret);
+		return;
+	}
+
+	/* convert stored data to params struct */
+	params.ssid = creds.header.ssid;
+	params.ssid_length = creds.header.ssid_len;
+	params.security = creds.header.type;
+	if (creds.header.type == WIFI_SECURITY_TYPE_PSK ||
+	    creds.header.type == WIFI_SECURITY_TYPE_PSK_SHA256) {
+		params.psk = creds.password;
+		params.psk_length = creds.password_len;
+	}
+	if (creds.header.type == WIFI_SECURITY_TYPE_SAE) {
+		params.sae_password = creds.password;
+		params.sae_password_length = creds.password_len;
+	}
+	params.mfp = 1; /* management frame protection: optional */
+	params.channel = WIFI_CHANNEL_ANY;
+	params.band = 255; /* use any band */
+
+	/* add network to wpa_supplicant's RAM storage */
+	add_network(wpa_s, &params);
+}
+
+/* } */
+
+static int wifi_ext_command(uint32_t mgmt_request, struct net_if *iface,
+			    void *data, size_t len)
+{
+	struct wpa_supplicant *wpa_s = NULL;
+	const struct device *dev = net_if_get_device(iface);
+	struct wifi_connect_req_params *params = { 0 };
+	int ret = 0;
+
+	k_mutex_lock(&wpa_supplicant_mutex, K_FOREVER);
+
+	wpa_s = get_wpa_s_handle(dev);
+	if (!wpa_s) {
+		goto out;
+	}
+	switch (mgmt_request) {
+	case NET_REQUEST_WIFI_CONNECT_ANY:
+		wpas_request_connection(wpa_s);
```

Comment:
Please note that we are moving from directly calling wpa_supplicant APIs as this is prohibited as it violates the concurrency model of wpa_supplicant. But this is still a WIP, so, we can live with this.

Reply:
What would be the alternative to call wpa_supplicant APIs?

Reply:
@maxd-nordic Take a look here: https://github.com/nrfconnect/sdk-hostap/blob/main/zephyr/src/supp_api.h. The API is implemented with mutex. However, looking at this I feel that this might need further changes. While the calls to supplicant are protected there might be a race between supplicant thread and the external thread calling these functions. @krish2718 will know best.